File: Data Structures & Algorithms/delete-node-in-a-bst/submission-0.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* deleteNode(TreeNode* root, int key) {
        if (root == nullptr) {
            return root;
        }

        if (root->val > key) {
            // traverse to left 
            root->left = deleteNode(root->left, key);
        }
        else if (root->val < key) {
            // traverse to right
            root->right = deleteNode(root->right, key);
        }
        else {
            // we are on the node that needs to be deleted now
            if (root->left == nullptr) {
                return root->right;
            }
            else if (root->right == nullptr) {
                return root->left;
            } else {
                TreeNode *minNode = findMinValueNode(root->right);
                root->val = minNode->val;
                root->right = deleteNode(root->right, minNode->val);
            }
        }

        return root;
    }

    TreeNode* findMinValueNode(TreeNode *root) {
        TreeNode *curr = root;
        while (curr != nullptr && curr->left != nullptr) {
            curr = curr->left;
        }
        return curr;
    }



};

File: Data Structures & Algorithms/binary-tree-right-side-view/submission-0.rs
// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//     pub val: i32,
//     pub left: Option<Rc<RefCell<TreeNode>>>,
//     pub right: Option<Rc<RefCell<TreeNode>>>,
// }
//
// impl TreeNode {
//     #[inline]
//     pub fn new(val: i32) -> Self {
//         TreeNode {
//             val,
//             left: None,
//             right: None,
//         }
//     }
// }

use std::rc::Rc;
use std::cell::RefCell;

impl Solution {
    pub fn right_side_view(root: Option<Rc<RefCell<TreeNode>>>) -> Vec<i32> {

        if root.is_none() {
            return Vec::new();
        }

        let mut right_nodes = Vec::new();
        let mut dequeue: VecDeque<Rc<RefCell<TreeNode>>> = VecDeque::new();        
        let mut level = 0;

        let root_rc = root.unwrap();
        dequeue.push_front(root_rc.clone());
        // right_nodes.push(root_rc.borrow().val);

        while !dequeue.is_empty() {

            let mut curr_length = dequeue.len();

            for i in 0..dequeue.len() {

                // if curr index is one away from the len of the deque, 
                // this is the rightmost node and we should add it to the right_node 
                // vector

                let node_rc = dequeue.pop_front().unwrap();
                let node = node_rc.borrow();

                if curr_length - i <= 1 {
                    right_nodes.push(node.val);
                }

                if let Some(left) = node.left.clone() {
                    dequeue.push_back(left);
                }
                if let Some(right) = node.right.clone() {
                    dequeue.push_back(right); 
                }            

            }
        }

        return right_nodes;

    }
}
